lint/base_linter/node_linter.py

```diff
@@ -146,21 +147,67 @@ def find_local_executable(self, start_dir, npm_name):
                     self.notify_failure()
                     raise linter.PermanentError()
 
-                # A 'package.json' not yet installed?
-                if not os.path.exists(os.path.join(path, 'node_modules', '.bin')):
-                    is_dep = bool(manifest.get('dependencies', {}).get(npm_name))
-                    is_dev_dep = bool(manifest.get('devDependencies', {}).get(npm_name))
-                    if is_dep or is_dev_dep:
-                        logger.warning(
-                            "Skipping '{}' for now which is listed as a {} "
-                            "in {} but not installed.  Forgot to 'npm install'?"
-                            .format(
-                                npm_name,
-                                'dependency' if is_dep else 'devDependency',
-                                manifest_file
+                is_dep = bool(manifest.get('dependencies', {}).get(npm_name))
+                is_dev_dep = bool(manifest.get('devDependencies', {}).get(npm_name))
+                if is_dep or is_dev_dep:
+                    self.context['project_root'] = path
+
+                    # Perhaps this is a Yarn PnP project?
+                    yarn_path = shutil.which('yarn')
```

Comment:
which yarn is way more expensive than the other checks, so we should think about flipping the conditionals. I.e. on Windows you often have lots of paths in PATH, which does an fs.stat for every path whereby checking if pnp.js exists is just one fs.stat call.

Reply:
I noticed there's a `self.which` method. Is that more optimized or does it just delegate to `shutil.which`?

Reply:
Yeah, typically

	if is_yarn_pnp_project:
		yarn_path = ...
		if not yarn_path:
			raise ...
		return [yarn, run, ...]

no?